i6_models/parts/conformer/mhsa_rel_pos.py

```diff
@@ -0,0 +1,248 @@
+
+        if self.learnable_pos_emb:
+            pos_seq_q = torch.arange(time_dim_size, device=input_tensor.device)
+            pos_seq_k = torch.arange(time_dim_size, device=input_tensor.device)
+
+            distance_mat = pos_seq_k[None, :] - pos_seq_q[:, None]
+            distance_mat_clipped = torch.clamp(distance_mat, -self.rel_pos_clip, self.rel_pos_clip)
+
+            final_mat = distance_mat_clipped + self.rel_pos_clip
+
+            rel_pos_embeddings = self.rel_pos_embeddings[final_mat]  # [T, T', pos_emb_dim]
+        else:
+            rel_pos_embeddings = self._sinusoidal_pe(
+                torch.arange(time_dim_size - 1, -time_dim_size, -1, device=input_tensor.device, dtype=torch.float32),
+                self.pos_emb_dim,
+            ).view(
+                1, 2 * time_dim_size - 1, self.pos_emb_dim
+            )  # [1, T+T'-1, pos_emb_dim]
+
+        # dropout relative positional embeddings
+        rel_pos_embeddings = self.pos_emb_dropout(
+            rel_pos_embeddings
+        )  # [T, T', pos_emb_dim] or [1, T+T'-1, pos_emb_dim]
+        rel_pos_embeddings = rel_pos_embeddings.unsqueeze(2)  # [T, T', 1, pos_emb_dim] or [1, T+T'-1, 1, pos_emb_dim]
+
+        # linear transformation or identity
+        rel_pos_embeddings = self.linear_pos(rel_pos_embeddings)  # [T, T', 1, F'|F] or [1, T+T'-1, 1, F'|F]
+
+        if self.separate_pos_emb_per_head:
+            rel_pos_embeddings = rel_pos_embeddings.squeeze(2).reshape(
+                *rel_pos_embeddings.shape[:2], -1, self.embed_dim_per_head
+            )  # [T, T', #heads, F'] or [1, T+T'-1, #heads, F']
+
+        q_with_bias_u = q + self.pos_bias_u if self.with_pos_bias else q  # [B, T, #heads, F']
+        q_with_bias_v = q + self.pos_bias_v if self.with_pos_bias else q
+
+        # attention matrix a and c
+        attn_ac = torch.einsum("bihf, bjhf -> bhij", q_with_bias_u, k)  # [B, #heads, T, T']
+
+        # attention matrix b and d
+        attn_bd = torch.einsum(
+            "bihf, ijhf -> bhij", q_with_bias_v, rel_pos_embeddings
+        )  # [B, #heads, T, T'] or [B, #heads, T, T+T'+1]
+
+        if not self.learnable_pos_emb:
+            attn_bd = self._rel_shift_bhij(attn_bd, k_len=time_dim_size)  # [B, #heads, T, T']
+
+        attn = attn_ac + attn_bd + mask  # [B, #heads, T, T']
+        attn_scaled = attn * (math.sqrt(1.0 / float(self.embed_dim_per_head)))  # [B, #heads, T, T']
+
+        # softmax and dropout
+        attn_output_weights = self.att_weights_dropout(F.softmax(attn_scaled, dim=-1))  # [B, #heads, T, T']
+
+        # sequence of weighted sums over value sequence
+        v = value_seq.view(batch_dim_size, -1, self.num_heads, self.embed_dim_per_head)  # [B, T, H, F']
+        attn_output = (
+            torch.einsum("bhij, bjhf -> bhif", attn_output_weights, v)
+            .transpose(1, 2)
+            .contiguous()
+            .view(batch_dim_size, -1, self.embed_dim)
+        )
```

Comment:
I don't understand:
* Why `contiguous` and then `view`? There is `reshape` exactly for that.
* Why the `transpose` when you already have `einsum` before?

So, shouldn't this be the same but just simpler?
```suggestion
        attn_output = (
            torch.einsum("bhij, bjhf -> bihf", attn_output_weights, v)
            .reshape(batch_dim_size, -1, self.embed_dim)
        )
```